esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -6118,7 +6104,289 @@ bool CWsDfuEx::onDFUFileAccess(IEspContext &context, IEspDFUFileAccessRequest &r
+        if (isEmptyString(fileId))
+             throw MakeStringException(ECLWATCH_INVALID_INPUT, "No FileId defined.");
+        if (isEmptyString(recordDefinition))
+             throw MakeStringException(ECLWATCH_INVALID_INPUT, "No ECLRecordDefinition defined.");
+
+        StringBuffer userId, oldFileName, newFileName, clusterName, groupName;
+        const char *pGroupName = strchr(fileId, DFUFileIdSeparator);
+        if (!pGroupName)
+            throw MakeStringException(ECLWATCH_INVALID_INPUT, "Failed to read Group Name from FileId '%s'.", fileId);
+
+        groupName.append(pGroupName - fileId, fileId);
+        const char *pClusterName = strchr(++pGroupName, DFUFileIdSeparator);
+        if (!pClusterName)
+            throw MakeStringException(ECLWATCH_INVALID_INPUT, "Failed to read Cluster Name from FileId '%s'.", fileId);
+        clusterName.append(pClusterName - pGroupName, pGroupName);
+
+        oldFileName.set(++pClusterName);
+        newFileName.set(pClusterName);
+        if (newFileName.length() < 17)
+            throw MakeStringException(ECLWATCH_INVALID_INPUT, "Invalid FileId: cannot read FileName.");
+        newFileName.setLength(newFileName.length() - 16); //".wsdfucreate.tmp"
+
+        Owned<IConstWUClusterInfo> clusterInfo = getTargetClusterInfo(clusterName.str());
+        ClusterType clusterType = clusterInfo->getPlatform();
+        if (clusterType == RoxieCluster)
+            throw MakeStringException(ECLWATCH_INVALID_INPUT, "Roxie cluster is not allowed: %s.", clusterName.str());
+
+        MemoryBuffer layoutBin;
+        exportRecordDefinitionBinaryType(recordDefinition, layoutBin);
+
+        Owned<IUserDescriptor> userDesc;
+        context.getUserID(userId);
+        if (!userId.isEmpty())
+        {
+            userDesc.setown(createUserDescriptor());
+            userDesc->set(userId.str(), context.queryPassword(), context.querySignature());
+        }
+
+        VStringBuffer jobId("Publish %s on %s", newFileName.str(), clusterName.str());
+        ESPLOG(LogMax, "%s", jobId.str());
+        StringArray dummyArray;
+        Owned<IFileDescriptor> fileDesc = createDFUDescriptor(context, oldFileName.str(),
+            clusterType, jobId.str(), userId.str(), dummyArray, groupName);
+        Owned<IDistributedFile> newFile = queryDistributedFileDirectory().createNew(fileDesc);
+        newFile->validate();
+
+        //Jake, please check whether the renamePhysicalPartFiles() is called correctly or not.
+        if (!newFile->renamePhysicalPartFiles(newFileName.str(), nullptr, nullptr, nullptr))
+            throw MakeStringException(ECLWATCH_FILE_NOT_EXIST, "Failed in renamePhysicalPartFiles %s.", newFileName.str());
+
+        newFile->setAccessed();
+        newFile->setECL(recordDefinition);
+        newFile->queryAttributes().setPropBin("_rtlType", layoutBin.length(), layoutBin.toByteArray());
+        newFile->queryAttributes().setPropInt64("@recordCount", req.getRecordCount());
```

Comment:
What if there is no record count? Should avoid setting to 0 if so.

Reply:
Ok